test/test_response.py

```diff
@@ -448,6 +448,45 @@ def test_decode_zstd_incomplete(self, data: bytes) -> None:
         with pytest.raises(DecodeError):
             HTTPResponse(fp, headers={"content-encoding": "zstd"})
 
+    decode_param_set = [
+        b"f",
+        b"fo",
+        b"foo",
+        b"foo" * 5,
+        b"123456789012345",
+        b"foo4",
+        b"x" * 100,
+    ]
+
+    @onlyZstd()
+    @pytest.mark.parametrize("data", decode_param_set)
+    def test_decode_zstd_incomplete_read(self, data: bytes) -> None:
+        data = zstd.compress(data)
+        fp = BytesIO(data[:-1])  # shorten the data to trigger DecodeError
+
+        # create response object without(!) reading/decoding the content
+        r = HTTPResponse(
+            fp, headers={"content-encoding": "zstd"}, preload_content=False
+        )
+
+        # read/decode, expecting DecodeError
+        with pytest.raises(DecodeError):
+            r.read(decode_content=True)
+
+    @onlyZstd()
+    @pytest.mark.parametrize("data", decode_param_set)
+    def test_decode_zstd_incomplete_read1(self, data: bytes) -> None:
+        data = zstd.compress(data)
+        fp = BytesIO(data[:-1])
+
+        r = HTTPResponse(
+            fp, headers={"content-encoding": "zstd"}, preload_content=False
+        )
+
+        # read/decode via read1(!), expecting DecodeError
+        with pytest.raises(DecodeError):
+            r.read1(decode_content=True)
```

Comment:
Thanks for checking the new `read1` functionality!

In this case, I am not convinced there is a bug, the tests look incorrect. You can get this test to stop failing if you add one more `r.read1(decode_content=True)` call. Similarly, you can make `test_decode_zstd_incomplete_read` fail if you call `r.read(amt=len(data), decode_content=True)`.

`read1` is not guaranteed to read all the data. If it reads a piece of the compressed data and successfully decodes the data, it returns it without flushing the decoder. This is just like `read` when passed `amt` is less than the total content length.

In this case, `r.read1(decode_content=True)` returns `data[:-1]`.

Reply:
I think i've understood.